sky/backends/provision_utils.py

```diff
@@ -0,0 +1,510 @@
+
+_MAX_RETRY = 3
+_TITLE = '\n\n' + '=' * 20 + ' {} ' + '=' * 20 + '\n'
+
+
+@dataclasses.dataclass
+class ClusterName:
+    display_name: str
+    name_on_cloud: str
+
+    def __repr__(self) -> str:
+        return repr(self.display_name)
+
+    def __str__(self) -> str:
+        return self.display_name
+
+
+@contextlib.contextmanager
+def _add_logger_handlers(log_path: str):
+    """Add file handler for logger."""
+    try:
+        log_abs_path = pathlib.Path(log_path).expanduser().absolute()
+        fh = logging.FileHandler(log_abs_path)
+        fh.setFormatter(sky_logging.FORMATTER)
+        fh.setLevel(logging.DEBUG)
+        logger.addHandler(fh)
+
+        # Redirect underlying provision logs to file.
+        provision_logger = logging.getLogger('sky.provision')
+        stream_handler = sky_logging.RichSafeStreamHandler(sys.stdout)
+        stream_handler.flush = sys.stdout.flush  # type: ignore
+        stream_handler.setFormatter(sky_logging.DIM_FORMATTER)
+        stream_handler.setLevel(logging.WARNING)
+
+        provision_logger.addHandler(fh)
+        provision_logger.addHandler(stream_handler)
+
+        provision_config.config.provision_log = log_abs_path
+        yield
+    finally:
+        logger.removeHandler(fh)
+        provision_logger.removeHandler(fh)
+        provision_logger.removeHandler(stream_handler)
+        stream_handler.close()
+        fh.close()
+
+
+def _bulk_provision(
+    cloud: clouds.Cloud,
+    region: clouds.Region,
+    zones: Optional[List[clouds.Zone]],
+    cluster_name: ClusterName,
+    bootstrap_config: provision_common.InstanceConfig,
+) -> provision_common.ProvisionMetadata:
+    provider_name = repr(cloud)
+    region_name = region.name
+
+    style = colorama.Style
+
+    if not zones:
+        # For Azure, zones is always an empty list.
+        zone_str = 'all zones'
+    else:
+        zone_str = ','.join(z.name for z in zones)
+
+    if isinstance(cloud, clouds.Local):
+        logger.info(f'{style.BRIGHT}Launching on local cluster '
+                    f'{cluster_name!r}.')
+    else:
+        logger.info(f'{style.BRIGHT}Launching on {cloud} '
+                    f'{region_name}{style.RESET_ALL} ({zone_str})')
+
+    start = time.time()
+    with rich_utils.safe_status('[bold cyan]Launching[/]') as status:
+        try:
+            # TODO(suquark): Should we cache the bootstrapped result?
+            #  Currently it is not necessary as bootstrapping takes
+            #  only ~3s, caching it seems over-engineering and could
+            #  cause other issues like the cache is not synced
+            #  with the cloud configuration.
+            config = provision.bootstrap_instances(provider_name, region_name,
+                                                   cluster_name.name_on_cloud,
+                                                   bootstrap_config)
+        except Exception as e:
+            logger.error('Failed to bootstrap configurations for '
```

Comment:
nit: Can we reword? Users would get confused by what "bootstrap configurations" means.

```
…
I 09-25 21:35:36 provision_utils.py:103] Launching on AWS us-east-2 (us-east-2c)
E 09-25 21:35:37 provision_utils.py:118] Failed to bootstrap configurations for 'dbg-private-vpc2'.
E 09-25 21:35:37 provision_utils.py:200] Failed to provision 'dbg-private-vpc2' on AWS (us-east-2c).
E 09-25 21:35:37 provision_utils.py:211] Terminating the failed cluster.
W 09-25 21:35:37 cloud_vm_ray_backend.py:2184] sky.exceptions.ResourcesUnavailableError: Failed to acquire resources in us-east-2c. Try changing resource requirements or use another zone.
W 09-25 21:35:37 cloud_vm_ray_backend.py:2193]
W 09-25 21:35:37 cloud_vm_ray_backend.py:2193] Provision failed for 1x AWS(t3.micro[Spot]) in us-east-2c. Trying other locations (if any).
```